File: discord-bot/mcp-server/src/tools/scanner.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { prisma } from "../db.js";
import { getItemDisplayName } from "../utils/items.js";

const SCANNER_URL = process.env.SCANNER_URL || "http://localhost:8001";

interface ScannerResult {
  items: Array<{
    code: string;
    quantity: number;
    crated: boolean;
    confidence: number;
  }>;
  faction?: string;
  error?: string;
}

export function registerScannerTools(server: McpServer): void {
  // scan_screenshot - Process a stockpile screenshot via OCR
  server.tool(
    "scan_screenshot",
    "Process a stockpile screenshot via OCR to extract item inventory",
    {
      imageUrl: z.string().url().describe("URL of the screenshot to process"),
      faction: z.enum(["colonials", "wardens", "all"]).optional().default("all"),
    },
    async ({ imageUrl, faction }) => {
      try {
        // Download the image
        const imageResponse = await fetch(imageUrl);
        if (!imageResponse.ok) {
          return {
            content: [
              { type: "text" as const, text: JSON.stringify({ error: "Failed to download image" }) },
            ],
            isError: true,
          };
        }

        const imageBuffer = await imageResponse.arrayBuffer();
        const imageBase64 = Buffer.from(imageBuffer).toString("base64");

        // Send to scanner service
        const scanResponse = await fetch(`${SCANNER_URL}/scan`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            image: imageBase64,
            faction: faction || "all",
          }),
        });

        if (!scanResponse.ok) {
          return {
            content: [
              {
                type: "text" as const,
                text: JSON.stringify({
                  error: "Scanner service error",
                  status: scanResponse.status,
                }),
              },
            ],
            isError: true,
          };
        }

        const result: ScannerResult = await scanResponse.json();

        if (result.error) {
          return {
            content: [
              { type: "text" as const, text: JSON.stringify({ error: result.error }) },
            ],
            isError: true,
          };
        }

        // Format results with display names
        const items = result.items.map((item) => ({
          itemCode: item.code,
          displayName: getItemDisplayName(item.code),
          quantity: item.quantity,
          crated: item.crated,
          confidence: Math.round(item.confidence * 100),
        }));

        // Sort by quantity descending
        items.sort((a, b) => b.quantity - a.quantity);

        const totalItems = items.reduce((sum, item) => sum + item.quantity, 0);
        const avgConfidence = items.length > 0
          ? Math.round(items.reduce((sum, item) => sum + item.confidence, 0) / items.length)
          : 0;

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({
                success: true,
                itemCount: items.length,
                totalQuantity: totalItems,
                averageConfidence: avgConfidence,
                faction: result.faction || faction,
                items,
                note: "Use save_scan_results to save these items to a stockpile",
              }, null, 2),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({
                error: "Failed to process screenshot",
                details: error instanceof Error ? error.message : "Unknown error",
              }),
            },
          ],
          isError: true,
        };
      }
    }
  );

  // save_scan_results - Save scanned items to a stockpile
  server.tool(
    "save_scan_results",
    "Save OCR scan results to a stockpile",
    {
      regimentId: z.string().describe("Discord guild ID of the regiment"),
      userId: z.string().describe("User ID saving the scan"),
      stockpileId: z.string().describe("Stockpile ID to save items to"),
      items: z.array(z.object({
        itemCode: z.string(),
        quantity: z.number().min(0),
        crated: z.boolean().optional().default(false),
        confidence: z.number().min(0).max(100).optional(),
      })).min(1),
    },
    async ({ regimentId, userId, stockpileId, items }) => {
      // Verify stockpile belongs to regiment
      const stockpile = await prisma.stockpile.findFirst({
        where: { id: stockpileId, regimentId },
      });

      if (!stockpile) {
        return {
          content: [
            { type: "text" as const, text: JSON.stringify({ error: "Stockpile not found" }) },
          ],
          isError: true,
        };
      }

      // Create scan record and upsert items in transaction
      const result = await prisma.$transaction(async (tx) => {
        // Create scan record
        const scan = await tx.stockpileScan.create({
          data: {
            stockpileId,
            scannedById: userId,
            itemCount: items.length,
            ocrConfidence: items.reduce((sum, i) => sum + (i.confidence || 0), 0) / items.length / 100,
          },
        });

        // Create scan items for history
        await tx.stockpileScanItem.createMany({
          data: items.map((item) => ({
            scanId: scan.id,
            itemCode: item.itemCode,
            quantity: item.quantity,
            crated: item.crated || false,
            confidence: item.confidence ? item.confidence / 100 : null,
          })),
        });

        // Upsert stockpile items (replace existing quantities)
        for (const item of items) {
          await tx.stockpileItem.upsert({
            where: {
              stockpileId_itemCode_crated: {
                stockpileId,
                itemCode: item.itemCode,
                crated: item.crated || false,
              },
            },
            create: {
              stockpileId,
              itemCode: item.itemCode,
              quantity: item.quantity,
              crated: item.crated || false,
              confidence: item.confidence ? item.confidence / 100 : null,
            },
            update: {
              quantity: item.quantity,
              confidence: item.confidence ? item.confidence / 100 : null,
            },
          });
        }

        // Update stockpile timestamp
        await tx.stockpile.update({
          where: { id: stockpileId },
          data: { updatedAt: new Date() },
        });

        return scan;
      });

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({
              success: true,
              scanId: result.id,
              stockpileId,
              stockpileName: stockpile.name,
              itemsSaved: items.length,
              totalQuantity: items.reduce((sum, i) => sum + i.quantity, 0),
            }, null, 2),
          },
        ],
      };
    }
  );
}

File: mcp-server/src/tools/scanner.py
"""Scanner MCP tools - OCR processing for stockpile screenshots."""

import json
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Any

import httpx
from sqlalchemy import delete, select

from ..config import settings
from ..db.session import get_session
from ..db.models import Stockpile, StockpileItem, StockpileScan, StockpileScanItem
from ..db.models.regiment import Regiment
from ..db.models.user import User
from ..foxhole.items import get_item_display_name
from ..utils.logger import logger

if TYPE_CHECKING:
    from ..server import McpServer


def register_scanner_tools(server: "McpServer") -> None:
    """Register scanner-related MCP tools."""

    async def scan_screenshot(args: dict[str, Any]) -> dict[str, Any]:
        """Process a stockpile screenshot via OCR to extract item inventory."""
        image_url = args["imageUrl"]
        faction = args.get("faction", "all")

        try:
            async with httpx.AsyncClient(timeout=60.0) as client:
                # Download the image
                image_response = await client.get(image_url)
                if image_response.status_code != 200:
                    return {
                        "content": [
                            {"type": "text", "text": json.dumps({"error": "Failed to download image"})},
                        ],
                        "isError": True,
                    }

                # Send to scanner service using multipart form (matches web app)
                files = {"image": ("screenshot.png", image_response.content, "image/png")}
                data = {"faction": faction}
                scan_response = await client.post(
                    f"{settings.scanner_url}/ocr/scan_image",
                    files=files,
                    data=data,
                )

                if scan_response.status_code != 200:
                    return {
                        "content": [
                            {
                                "type": "text",
                                "text": json.dumps({
                                    "error": "Scanner service error",
                                    "status": scan_response.status_code,
                                }),
                            },
                        ],
                        "isError": True,
                    }

                result = scan_response.json()

                if result.get("error"):
                    return {
                        "content": [
                            {"type": "text", "text": json.dumps({"error": result["error"]})},
                        ],
                        "isError": True,
                    }

                # Extract detected stockpile name from scanner response
                detected_name = (
                    result.get("stockpileName")
                    or result.get("stockpile_name")
                    or result.get("name")
                )
                detected_type = result.get("stockpileType") or result.get("stockpile_type")

                # Format results with display names
                items = [
                    {
                        "itemCode": item["code"],
                        "displayName": get_item_display_name(item["code"]),
                        "quantity": item["quantity"],
                        "crated": item["crated"],
                        "confidence": round(item["confidence"] * 100),
                    }
                    for item in result.get("items", [])
                ]

                # Sort by quantity descending
                items.sort(key=lambda x: x["quantity"], reverse=True)

                total_items = sum(item["quantity"] for item in items)
                avg_confidence = (
                    round(sum(item["confidence"] for item in items) / len(items))
                    if items
                    else 0
                )

                return {
                    "content": [
                        {
                            "type": "text",
                            "text": json.dumps({
                                "success": True,
                                "itemCount": len(items),
                                "totalQuantity": total_items,
                                "averageConfidence": avg_confidence,
                                "faction": result.get("faction", faction),
                                "detectedName": detected_name,
                                "detectedType": detected_type,
                                "items": items,
                                "note": "Use save_scan_results to save these items to a stockpile",
                            }, indent=2),
                        },
                    ],
                }

        except httpx.TimeoutException:
            return {
                "content": [
                    {"type": "text", "text": json.dumps({"error": "Scanner service timeout"})},
                ],
                "isError": True,
            }
        except Exception as e:
            logger.error(f"Scanner error: {e}")
            return {
                "content": [
                    {
                        "type": "text",
                        "text": json.dumps({
                            "error": "Failed to process screenshot",
                            "details": str(e),
                        }),
                    },
                ],
                "isError": True,
            }

    server.tool(
        "scan_screenshot",
        "Process a stockpile screenshot via OCR to extract item inventory",
        {
            "imageUrl": {
                "type": "string",
                "description": "URL of the screenshot to process",
                "required": True,
            },
            "faction": {
                "type": "string",
                "description": "Faction filter: colonials, wardens, all",
                "default": "all",
            },
        },
        scan_screenshot,
    )

    async def save_scan_results(args: dict[str, Any]) -> dict[str, Any]:
        """Save OCR scan results to a stockpile."""
        regiment_id = args["regimentId"]
        user_id = args["userId"]
        stockpile_id = args["stockpileId"]
        items = args["items"]

        async with get_session() as session:
            # Verify stockpile belongs to regiment
            stockpile_result = await session.execute(
                select(Stockpile)
                .where(Stockpile.id == stockpile_id)
                .where(Stockpile.regimentId == regiment_id)
            )
            stockpile = stockpile_result.scalar()

            if not stockpile:
                return {
                    "content": [
                        {"type": "text", "text": json.dumps({"error": "Stockpile not found"})},
                    ],
                    "isError": True,
                }

            # Calculate average confidence
            avg_confidence = (
                sum(item.get("confidence", 0) for item in items) / len(items) / 100
                if items
                else 0
            )

            # Create scan record
            scan = StockpileScan(
                id=str(uuid.uuid4()),
                stockpileId=stockpile_id,
                scannedById=user_id,
                itemCount=len(items),
                ocrConfidence=avg_confidence,
            )
            session.add(scan)

            # Create scan items for history
            for item in items:
                scan_item = StockpileScanItem(
                    id=str(uuid.uuid4()),
                    scanId=scan.id,
                    itemCode=item["itemCode"],
                    quantity=item["quantity"],
                    crated=item.get("crated", False),
                    confidence=item.get("confidence", 0) / 100 if item.get("confidence") else None,
                )
                session.add(scan_item)

            # Delete all existing items then create new ones (matches web app behavior)
            await session.execute(
                delete(StockpileItem)
                .where(StockpileItem.stockpileId == stockpile_id)
            )

            # Create new items from scan
            for item in items:
                new_item = StockpileItem(
                    id=str(uuid.uuid4()),
                    stockpileId=stockpile_id,
                    itemCode=item["itemCode"],
                    quantity=item["quantity"],
                    crated=item.get("crated", False),
                    confidence=item.get("confidence", 0) / 100 if item.get("confidence") else None,
                )
                session.add(new_item)

            # Update stockpile timestamp
            stockpile.updatedAt = datetime.utcnow()

            await session.commit()

        return {
            "content": [
                {
                    "type": "text",
                    "text": json.dumps({
                        "success": True,
                        "scanId": scan.id,
                        "stockpileId": stockpile_id,
                        "stockpileName": stockpile.name,
                        "itemsSaved": len(items),
                        "totalQuantity": sum(item["quantity"] for item in items),
                    }, indent=2),
                },
            ],
        }

    server.tool(
        "save_scan_results",
        "Save OCR scan results to a stockpile",
        {
            "regimentId": {
                "type": "string",
                "description": "Discord guild ID of the regiment",
                "required": True,
            },
            "userId": {
                "type": "string",
                "description": "User ID saving the scan",
                "required": True,
            },
            "stockpileId": {
                "type": "string",
                "description": "Stockpile ID to save items to",
                "required": True,
            },
            "items": {
                "type": "array",
                "description": "Items to save",
                "required": True,
            },
        },
        save_scan_results,
    )

    async def get_scanner_channel(args: dict[str, Any]) -> dict[str, Any]:
        """Get the configured scanner channel for a regiment."""
        regiment_id = args["regimentId"]

        async with get_session() as session:
            result = await session.execute(
                select(Regiment.scannerChannelId)
                .where(Regiment.discordId == regiment_id)
            )
            row = result.scalar()

            return {
                "content": [
                    {
                        "type": "text",
                        "text": json.dumps({
                            "scannerChannelId": row,
                        }),
                    },
                ],
            }

    server.tool(
        "get_scanner_channel",
        "Get the configured scanner channel ID for a regiment",
        {
            "regimentId": {
                "type": "string",
                "description": "Discord guild ID of the regiment",
                "required": True,
            },
        },
        get_scanner_channel,
    )

    async def resolve_discord_user(args: dict[str, Any]) -> dict[str, Any]:
        """Resolve a Discord user ID to an internal user ID."""
        discord_id = args["discordId"]

        async with get_session() as session:
            result = await session.execute(
                select(User.id)
                .where(User.discordId == discord_id)
            )
            user_id = result.scalar()

            if not user_id:
                return {
                    "content": [
                        {
                            "type": "text",
                            "text": json.dumps({
                                "error": "User not found",
                                "message": "This Discord user has not signed in to Foxhole Quartermaster yet.",
                            }),
                        },
                    ],
                    "isError": True,
                }

            return {
                "content": [
                    {
                        "type": "text",
                        "text": json.dumps({
                            "userId": user_id,
                        }),
                    },
                ],
            }

    server.tool(
        "resolve_discord_user",
        "Resolve a Discord user ID to an internal user ID",
        {
            "discordId": {
                "type": "string",
                "description": "Discord user ID to look up",
                "required": True,
            },
        },
        resolve_discord_user,
    )
